amads/melody/contour/interpolation_contour.py

```diff
@@ -0,0 +1,272 @@
+
+    @staticmethod
+    def _find_contour_extrema(pitches: list[int]) -> list[int]:
+        """Determine contour extremum notes (local minima and maxima, including endpoints).
+
+        Excludes changing notes (notae cambiatae).
+
+        Parameters
+        ----------
+        pitches : list[int]
+            List of MIDI pitch values
+
+        Returns
+        -------
+        numpy.ndarray
+            Array of indices corresponding to extrema points in the melody
+        """
+        extrema_indices = []
+        n = len(pitches)
+
+        # Always include first and last notes
+        extrema_indices.append(0)
+
+        # Find local extrema
+        for i in range(1, n - 1):
+            # Check previous and next notes
+            prev_pitch = pitches[i - 1]
+            curr_pitch = pitches[i]
+            next_pitch = pitches[i + 1]
+
+            # Skip changing notes (where adjacent pitches are equal)
+            if prev_pitch == next_pitch:
+                continue
+
+            # Check if it's a peak or valley
+            if (prev_pitch < curr_pitch and next_pitch < curr_pitch) or (
+                prev_pitch > curr_pitch and next_pitch > curr_pitch
+            ):
+                # Additional check for equal adjacent notes
+                is_extremum = True
+                if i > 1 and prev_pitch == curr_pitch:
+                    is_extremum = (pitches[i - 2] - curr_pitch) * (
+                        next_pitch - curr_pitch
+                    ) < 0
+                if i < n - 2 and next_pitch == curr_pitch:
+                    is_extremum = (prev_pitch - curr_pitch) * (
+                        pitches[i + 2] - curr_pitch
+                    ) < 0
+                if is_extremum:
+                    extrema_indices.append(i)
+
+        extrema_indices.append(n - 1)  # Add last note
+        return np.array(extrema_indices)
+
+    def calculate_interpolation_contour(self) -> list[float]:
+        """Calculate the interpolation contour [1] representation of a melody.
+
+        Returns
+        -------
+        list[float]
+            Array containing the interpolation contour representation
+
+        References
+        ----------
+        [1] Müllensiefen, D. (2009). FANTASTIC: Feature ANalysis Technology Accessing
+        STatistics (In a Corpus): Technical Report v1.5
+        https://www.doc.gold.ac.uk/isms/m4s/FANTASTIC_docs.pdf
+        """
+        # Step 1 & 2: Find contour extrema
+        extrema_indices = self._find_contour_extrema(self.pitches)
+        extrema_times = [self.times[i] for i in extrema_indices]
+        extrema_pitches = [self.pitches[i] for i in extrema_indices]
+
+        # Step 3: Calculate gradients
+        gradients = np.diff(extrema_pitches) / np.diff(extrema_times)
+
+        # Step 4: Calculate durations
+        durations = np.diff(extrema_times)
+
+        # Step 5: Remove durations below 0.05 seconds
+        durations[durations < 0.05] = 0
```

Comment:
0.05 s should be a parameter, not hard-coded

Reply:
In the more accurate ported version, this is handled by multiplying by 10 with a rounding - this is exactly how it's implemented in FANTASTIC, but i'm not opposed to extending the functionality per se